feat: add support for prometheus operator config

douglas-reid opened the pull request:
This PR is an attempt to add support for generating Prometheus Operator config to Istio. It is a port of https://github.com/istio/istio/pull/13407.

This PR assumes that the Prometheus Operator itself has been previously installed (example: `kubectl -n prom-operator apply -f https://raw.githubusercontent.com/coreos/prometheus-operator/master/bundle.yaml`).

Motivations (Code Mauve):
- https://github.com/istio/istio/issues/10134
- https://github.com/istio/istio/issues/7528
- https://github.com/istio/istio/issues/5673
- https://github.com/istio/istio/issues/12915
- https://discuss.istio.io/t/istio-prometheus-to-scrape-metrics-from-service-specific-metrics-endpoint/926
- https://discuss.istio.io/t/how-to-update-prometheus-config/935
- https://discuss.istio.io/t/istio-1-1-daily-byo-prometheus-via-operator/418
- https://discuss.istio.io/t/is-envoy-metrics-not-pushed-via-statsd-after-1-1/1883

NOTE: I have not tested this fully yet. I have only copied it over from the other PR.

costinm commented on istio-telemetry/prometheus-operator/values.yaml:
```diff
@@ -0,0 +1,18 @@
+# Controls the default scrape interval used in the ServiceMonitors
```
Please add a top-level "prometheus:", for consistency and compat with 1.1.
Can you do a diff with the values from 1.1, and maybe add a doc to any setting that is new, and 
a commented out setting if it was deleted ? 
( we need to do this for all components )

douglas-reid commented:
Note to self: need to update `istio-component-monitor` service monitor config, as some add/or all components may no longer be using a `http-monitoring` named port.

istio-testing commented:
[APPROVALNOTIFIER] This PR is **NOT APPROVED**

This pull-request has been approved by:
To fully approve this pull request, please assign additional approvers.
We suggest the following additional approvers: 

If they are not already assigned, you can assign the PR to them by writing `/assign ` in a comment when ready.

The full list of commands accepted by this bot can be found [here](https://go.k8s.io/bot-commands).

The pull request process is described [here](https://git.k8s.io/community/contributors/guide/owners.md#the-code-review-process)

<details open>
Needs approval from an approver in each of these files:

- **[OWNERS](https://github.com/istio/installer/blob/master/OWNERS)**

Approvers can indicate their approval by writing `/approve` in a comment
Approvers can cancel approval by writing `/approve cancel` in a comment
</details>
<!-- META={"approvers":[]} -->

douglas-reid commented:
PTAL.  I've moved the `values.yaml` options under a top-level `prometheus` tag, as requested.

I added some very high-level "tests" to the Makefile to validate the right config is generated, etc. I'm not sure if that is what is being sought, however.

douglas-reid commented:
FWIW, as far as I can tell, none of the test failures are related to these changes.

douglas-reid commented:
Friendly ping on review.

howardjohn commented on Makefile:
```diff
@@ -235,6 +235,19 @@ install-policy:
+# It is provided as a way to install Istio prometheus operator config in isolation.
+install-prometheus-operator-config:
+	kubectl create ns ${ISTIO_NS} || /bin/true
+	# NOTE: we don't use iop to install, as it defaults to `--prune`, which is incompatible with the prom operator (it prunes the stateful set)
```
How come? Does the statefulset not get the labels?

douglas-reid replied:
yeah, the operator doesn't apply the labels to the stateful set, it turns out.

howardjohn commented on Makefile:
```diff
@@ -235,6 +235,19 @@ install-policy:
+# This target expects that the prometheus operator (and its CRDs have already been installed).
+# It is provided as a way to install Istio prometheus operator config in isolation.
+install-prometheus-operator-config:
+	kubectl create ns ${ISTIO_NS} || /bin/true
```
can you change to `true` instead of `/bin/true`? Needed for mac compatibility apparently

douglas-reid commented:
Awesome!  Thanks for the reviews and help..

davidham commented:
Hi! Sorry to comment on a closed PR, but is there any documentation on how to use this? I am adding Istio to a cluster that already has prometheus-operator installed, and I'd love to have these metrics and dashboards be able to be sent to my existing install. Thank you!

howardjohn commented:
@davidham the charts are at https://github.com/istio/installer/tree/master/istio-telemetry/prometheus-operator

So you can just `helm template`/`helm install` those and it should work. If not, please let us know

davidham commented:
Ooooooooh nice! So if I installed Istio per the docs, should I just update my install to be `prometheus.enabled = false` and then install the prom-operator you mentioned above? Can I still make use of the grafana dashboards that come with Istio?

howardjohn commented:
Yep, you might need to configure grafana to point to the right prometheus though since its probably not on the same address as grafana expects out of the box

Martinlcksybo commented:
What is the name of the repository and chart? What's the helm command I should run?

davidham commented:
@howardjohn So: I wanted to install the Istio grafana dashboards, but use them in my existing prometheus/grafana setup from a separate install of `stable/prometheus-operator`. I installed the prom-operator from istio/installer/telemetry, and it seems to have worked because it installed ServiceMonitors for all the Istio stuff, and heir metrics show up in Prometheus. So far so good!

I also installed Grafana from istio/installer/telemetry, and this install worked also (though I had to get `fix_datasources.sh` to work first). It installed all the dashboard configmaps, but they aren't being picked up by my Grafana. Post-install I added the `grafana_dashboard` label to each one, and I updated the istio-grafana configmap to point to my main prometheus service. 

I looked at the logs for the grafana-sc-dashboard container, and i can see it reading all the configmaps, but it doesn't seem to be reading the ones in istio-system. Is there anything else I have to do to get these dashboards to work?

davidham commented:
☝️ Got this to work, I had neglected to create a rolebinding for Prometheus in the `istio-system` namespace. Ince I did this it worked fine.

mabushey commented:
> So you can just `helm template`/`helm install` those and it should work. If not, please let us know

Is there a way to just get the Yaml? Helm is too complicated for me. I can easily use SED or VIM to modify a yaml file. I don't understand layering on massive levels of complexity on top of something so simple and elegant as yaml config.

douglas-reid commented:
@mabushey i don't know that we have copy of these yamls with all the default values populated, but generating it yourself should be straightforward, even without helm. 

you could go the location of the template files (already linked above), grab the yaml that you want, copy it to wherever, and sed/vim away. you'll need to substitute out all the template stuff (contained within '{{' and '}}') with the values you want, of course.

fwiw, bending helm to your will can definitely be frustrating, but, for this specific generation task, just running `helm template` might be the simplest approach in the short term.

mabushey commented:
@douglas-reid 

I only use `helm template`, but usually it complains that the chart is not gziped or it throws an error like `Error: render error in "prometheus/templates/tests/test-prometheus-connection.yaml": template: prometheus/templates/tests/test-prometheus-connection.yaml:1:14: executing "prometheus/templates/tests/test-prometheus-connection.yaml" at <.Values.global.enableHelmTest>: nil pointer evaluating interface {}.enableHelmTest`

douglas-reid commented:
that happens when you set the chart to the lower level chart. hooray, helm!

the "secret" is to use the top-level chart and then (ugh) `-x <template>` for each template in the underlying chart (which here might just be one, if not, then both).

haf-afa commented:
@douglas-reid 

> i don't know that we have copy of these yamls with all the default values populated

This would be very useful. It would allow us to use kustomize.

> you could go the location of the template files (already linked above), grab the yaml that you want, copy it to wherever, and sed/vim away. you'll need to substitute out all the template stuff (contained within {{ and }}) with the values you want, of course.

This what that looks like for prometheus-operator: https://gist.github.com/haf-afa/575ed9f546037ea99c5b9e97f75b9c83#file-kustomization-yaml It took 2 days to make it work just for the operator with a stand-alone configured Grafana. Istio does both Grafana and Prom-Op, so I expect another two days for that. I've spent half a day so far, and gotten the dashboards to show up with the kiwigrid-sidecar fetching of configmaps. But it's not enough; I also have to modify the Prometheus configuration to match that of this repo, so that the values are populated in the Prometheus TSDB instance.

@mabushey I'd be happy to collaborate with you on getting this stack set up with plain Kubernetes files, the Prom Operator, Grafana and Istio. It would seem we have overlapping use-cases. I tried urging those good folk to do the same here https://github.com/coreos/prometheus-operator/issues/2853 - see the screenshots I posted; there were 163 different variables in that helm chart that could be tweaked when I counted; not counting the CRD definitions that could be inserted.
